### PR TITLE
#790 edit task endpoint

### LINKED ISSUE #790: Edit Task Endpoint (title, notes, priority, deadlines)
### Description

In order for users to edit tasks, an edit task endpoint route needs to be implemented in the backend. This is so that users can edit the tasks in the database

### Prerequisites

- [x] #765 

### Proposed Solution

Create the backend endpoint to support the editing of tasks. Endpoint will. verify required fields and validate inputs, then update database objects and return the edited task Id. 

- [x] POST endpoint on the path string “/:taskId/edit”
- [x] Accepted payload has 
  - [x] Text content for the updated title ‘title’
  - [x] Text content for the updated notes ‘notes’ 
  - [x] Text content for the updated priority ‘priority’
  - [x] Date content for the updated deadline ‘deadline’
- Controller will do the following:
  - [x] Get the current user 
  - [x] Call service function 
  - [x] Retrieve the task id from the query 
- Service will do the following:
  - [x] validate the user has the correct permissions
  - [x] Update the task with the passed in parameters
  - [x] Return the updated task

- Testing 
  - [x] Create Unit tests for all Errors thrown 
  - [x] Create Unit Tests for successful use of the method

### PR DESCRIPTION
**Pull request by ibarra000**
This PR depends on https://github.com/Northeastern-Electric-Racing/FinishLine/issues/789 being merged.

## Changes
Created editTask endpoint

## Notes

I'm getting some errors in the test data.
I added in checks that make sure the title and notes are within the proper word limits.

## Test Cases

- Successful task edit
- Unsuccessful task edit (invalid permissions, task deleted, task doesn't exist, title word limit, notes word limit ) 

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [ ] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #790

### CODE REVIEW

File: src/backend/src/services/tasks.services.ts [hunk 1]
```diff
@@ -61,4 +61,23 @@ export default class TasksService {
...
+  static async editTask(user: User, taskId: string, title: string, notes: string, priority: Task_Priority, deadline: Date) {
+    if (user.role === Role.GUEST) throw new AccessDeniedException();
+
+    const originalTask = await prisma.task.findUnique({ where: { taskId } });
+    if (!originalTask) throw new NotFoundException('Task', taskId);
+    if (originalTask.dateDeleted) throw new HttpException(400, 'Cant edit a deleted Task!');
+
+    if (!isUnderWordCount(title, 15)) throw new HttpException(400, 'Title must be less than 15 words');
+
+    if (!isUnderWordCount(notes, 150)) throw new HttpException(400, 'Notes must be less than 250 words');
```
Comment (Peyton-McKee) [inline src/backend/src/services/tasks.services.ts:74]: These numbers dont match up, think its supposed to be 250

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
Great work! Can u also post screenshots of manual tests pls.

File: src/backend/tests/tasks.test.ts [hunk 2]
```diff
@@ -98,3 +100,70 @@ describe('Tasks', () => {
     });
   });
 });
+
+describe('Edit Tasks', () => {
+  afterEach(() => {
```
Comment (Peyton-McKee) [inline src/backend/tests/tasks.test.ts:105]: You can delete this function we do it at the top

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]

File: src/backend/src/routes/tasks.routes.ts [hunk 3]
```diff
@@ -19,4 +19,13 @@ tasksRouter.post(
   TasksController.createTask
 );
 
+tasksRouter.post(
+  '/:taskId/edit',
+  nonEmptyString(body('title')),
+  nonEmptyString(body('notes')),
+  body('deadline'),
```
Comment (Peyton-McKee) [inline src/backend/src/routes/tasks.routes.ts:26]: validate this as a date please

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
One tiny thing that i see

File: src/backend/src/services/tasks.services.ts [hunk 4]
```diff
@@ -63,6 +63,35 @@ export default class TasksService {
...
+   * @param user the user editing the task
+   * @param taskId the task that is being edited
+   * @param title the new title for the task
+   * @param notes the new notes for the task
+   * @param priority the new priority for the task
+   * @param deadline the new deadline for the task
+   * @returns the sucessfully edited task
+   */
+  static async editTask(user: User, taskId: string, title: string, notes: string, priority: Task_Priority, deadline: Date) {
+    if (user.role === Role.GUEST) throw new AccessDeniedException();
```
Comment (anthonybernardi) [inline src/backend/src/services/tasks.services.ts:77]: can we use the has edit task permissions function created by Parth pls

**Review by anthonybernardi — CHANGES_REQUESTED** [review]

**Pull request merged**